File: libraries/python/qubi_protocol/builders.py
"""Command builders for creating type-safe Qubi commands."""

from typing import Dict, Any, Optional, Union

from .types import (
    QubiCommand,
    ModuleType,
    ServoParams,
    PositionParams,
    EyesParams,
    ExpressionParams,
    MovementParams,
    LocationParams,
    Expression,
)
from .errors import QubiValidationError


class BaseCommandBuilder:
    """Base class for command builders."""
    
    def __init__(self, module_id: str, module_type: ModuleType):
        self.module_id = module_id
        self.module_type = module_type
    
    def _create_command(self, action: str, params: Dict[str, Any]) -> QubiCommand:
        """Create a command with the given action and parameters."""
        return {
            "module_id": self.module_id,
            "module_type": self.module_type,
            "action": action,
            "params": params,
        }


class ActuatorCommandBuilder(BaseCommandBuilder):
    """Builder for actuator module commands."""
    
    def __init__(self, module_id: str):
        super().__init__(module_id, "actuator")
    
    def set_servo(self, angle: int, speed: Optional[int] = None, 
                  easing: Optional[str] = None) -> QubiCommand:
        """Create a servo control command."""
        self._validate_servo_params(angle, speed, easing)
        
        params: Dict[str, Union[int, str]] = {"angle": angle}
        if speed is not None:
            params["speed"] = speed
        if easing is not None:
            params["easing"] = easing
        
        return self._create_command("set_servo", params)
    
    def set_position(self, x: float, y: float, z: float) -> QubiCommand:
        """Create a 3D position command."""
        self._validate_position_params(x, y, z)
        
        params = {"x": x, "y": y, "z": z}
        return self._create_command("set_position", params)
    
    def get_position(self) -> QubiCommand:
        """Create a position query command."""
        return self._create_command("get_position", {})
    
    def stop(self) -> QubiCommand:
        """Create a stop command."""
        return self._create_command("stop", {})
    
    def _validate_servo_params(self, angle: int, speed: Optional[int], 
                              easing: Optional[str]) -> None:
        """Validate servo command parameters."""
        if not isinstance(angle, int) or angle < 0 or angle > 180:
            raise QubiValidationError("Servo angle must be an integer between 0 and 180")
        
        if speed is not None and (not isinstance(speed, int) or speed < 0 or speed > 255):
            raise QubiValidationError("Servo speed must be an integer between 0 and 255")
        
        if easing is not None:
            valid_easing = ["linear", "ease-in", "ease-out"]
            if easing not in valid_easing:
                raise QubiValidationError(f"Invalid easing type: {easing}")
    
    def _validate_position_params(self, x: float, y: float, z: float) -> None:
        """Validate position command parameters."""
        for coord, name in [(x, "x"), (y, "y"), (z, "z")]:
            if not isinstance(coord, (int, float)) or not (-float('inf') < coord < float('inf')):
                raise QubiValidationError(f"Position coordinate {name} must be a finite number")


class DisplayCommandBuilder(BaseCommandBuilder):
    """Builder for display module commands."""
    
    def __init__(self, module_id: str):
        super().__init__(module_id, "display")
    
    def set_eyes(self, left_x: int, left_y: int, right_x: int, right_y: int,
                 blink: Optional[bool] = None) -> QubiCommand:
        """Create an eye position command."""
        self._validate_eye_position(left_x, left_y, "left")
        self._validate_eye_position(right_x, right_y, "right")
        
        params: Dict[str, Any] = {
            "left_eye": {"x": left_x, "y": left_y},
            "right_eye": {"x": right_x, "y": right_y},
        }
        
        if blink is not None:
            params["blink"] = blink
        
        return self._create_command("set_eyes", params)
    
    def set_expression(self, expression: Expression, 
                      intensity: Optional[int] = None) -> QubiCommand:
        """Create a facial expression command."""
        self._validate_expression_params(expression, intensity)
        
        params: Dict[str, Union[str, int]] = {"expression": expression}
        if intensity is not None:
            params["intensity"] = intensity
        
        return self._create_command("set_expression", params)
    
    def clear_display(self) -> QubiCommand:
        """Create a clear display command."""
        return self._create_command("clear_display", {})
    
    def set_brightness(self, brightness: int) -> QubiCommand:
        """Create a brightness control command."""
        if not isinstance(brightness, int) or brightness < 0 or brightness > 100:
            raise QubiValidationError("Brightness must be an integer between 0 and 100")
        
        return self._create_command("set_brightness", {"brightness": brightness})
    
    def _validate_eye_position(self, x: int, y: int, eye_name: str) -> None:
        """Validate eye position parameters."""
        if not isinstance(x, int) or x < 0:
            raise QubiValidationError(f"{eye_name} eye x coordinate must be a non-negative integer")
        if not isinstance(y, int) or y < 0:
            raise QubiValidationError(f"{eye_name} eye y coordinate must be a non-negative integer")
    
    def _validate_expression_params(self, expression: str, intensity: Optional[int]) -> None:
        """Validate expression command parameters."""
        valid_expressions = ["happy", "sad", "surprised", "neutral", "angry"]
        if expression not in valid_expressions:
            raise QubiValidationError(f"Invalid expression: {expression}")
        
        if intensity is not None and (not isinstance(intensity, int) or intensity < 0 or intensity > 100):
            raise QubiValidationError("Expression intensity must be an integer between 0 and 100")


class MobileCommandBuilder(BaseCommandBuilder):
    """Builder for mobile module commands."""
    
    def __init__(self, module_id: str):
        super().__init__(module_id, "mobile")
    
    def move(self, velocity: float, direction: float, 
             duration: Optional[float] = None) -> QubiCommand:
        """Create a movement command."""
        self._validate_movement_params(velocity, direction, duration)
        
        params: Dict[str, Union[float]] = {
            "velocity": velocity,
            "direction": direction,
        }
        
        if duration is not None:
            params["duration"] = duration
        
        return self._create_command("move", params)
    
    def set_location(self, x: float, y: float, 
                    heading: Optional[float] = None) -> QubiCommand:
        """Create a location setting command."""
        self._validate_location_params(x, y, heading)
        
        params: Dict[str, float] = {"x": x, "y": y}
        if heading is not None:
            params["heading"] = heading
        
        return self._create_command("set_location", params)
    
    def get_location(self) -> QubiCommand:
        """Create a location query command."""
        return self._create_command("get_location", {})
    
    def stop(self) -> QubiCommand:
        """Create a stop command."""
        return self._create_command("stop", {})
    
    def rotate(self, angle: float, speed: Optional[float] = None) -> QubiCommand:
        """Create a rotation command."""
        if not isinstance(angle, (int, float)) or not (-float('inf') < angle < float('inf')):
            raise QubiValidationError("Rotation angle must be a finite number")
        
        params: Dict[str, Union[float]] = {"angle": angle}
        if speed is not None:
            if not isinstance(speed, (int, float)) or speed < 0 or speed > 100:
                raise QubiValidationError("Rotation speed must be a number between 0 and 100")
            params["speed"] = speed
        
        return self._create_command("rotate", params)
    
    def _validate_movement_params(self, velocity: float, direction: float, 
                                 duration: Optional[float]) -> None:
        """Validate movement command parameters."""
        if not isinstance(velocity, (int, float)) or not (-float('inf') < velocity < float('inf')):
            raise QubiValidationError("Velocity must be a finite number")
        
        if not isinstance(direction, (int, float)) or not (-float('inf') < direction < float('inf')):
            raise QubiValidationError("Direction must be a finite number")
        
        if duration is not None and (not isinstance(duration, (int, float)) or duration <= 0):
            raise QubiValidationError("Duration must be a positive number")
    
    def _validate_location_params(self, x: float, y: float, 
                                 heading: Optional[float]) -> None:
        """Validate location command parameters."""
        if not isinstance(x, (int, float)) or not (-float('inf') < x < float('inf')):
            raise QubiValidationError("Location x coordinate must be a finite number")
        
        if not isinstance(y, (int, float)) or not (-float('inf') < y < float('inf')):
            raise QubiValidationError("Location y coordinate must be a finite number")
        
        if heading is not None and (not isinstance(heading, (int, float)) or not (-float('inf') < heading < float('inf'))):
            raise QubiValidationError("Heading must be a finite number")


class SensorCommandBuilder(BaseCommandBuilder):
    """Builder for sensor module commands."""
    
    def __init__(self, module_id: str):
        super().__init__(module_id, "sensor")
    
    def read(self, sensor_type: Optional[str] = None) -> QubiCommand:
        """Create a sensor reading command."""
        params = {}
        if sensor_type is not None:
            params["sensor_type"] = sensor_type
        
        return self._create_command("read", params)
    
    def start_streaming(self, sensor_type: str, interval: float) -> QubiCommand:
        """Create a sensor streaming start command."""
        if not isinstance(sensor_type, str) or not sensor_type:
            raise QubiValidationError("Sensor type must be a non-empty string")
        
        if not isinstance(interval, (int, float)) or interval <= 0:
            raise QubiValidationError("Streaming interval must be a positive number")
        
        params = {"sensor_type": sensor_type, "interval": interval}
        return self._create_command("start_streaming", params)
    
    def stop_streaming(self, sensor_type: Optional[str] = None) -> QubiCommand:
        """Create a sensor streaming stop command."""
        params = {}
        if sensor_type is not None:
            params["sensor_type"] = sensor_type
        
        return self._create_command("stop_streaming", params)
    
    def calibrate(self, sensor_type: str) -> QubiCommand:
        """Create a sensor calibration command."""
        if not isinstance(sensor_type, str) or not sensor_type:
            raise QubiValidationError("Sensor type must be a non-empty string")
        
        params = {"sensor_type": sensor_type}
        return self._create_command("calibrate", params)
    
    def get_status(self) -> QubiCommand:
        """Create a sensor status query command."""
        return self._create_command("get_status", {})


class CustomCommandBuilder(BaseCommandBuilder):
    """Builder for custom module commands."""
    
    def __init__(self, module_id: str):
        super().__init__(module_id, "custom")
    
    def command(self, action: str, params: Optional[Dict[str, Any]] = None) -> QubiCommand:
        """Create a custom command."""
        if not isinstance(action, str) or not action:
            raise QubiValidationError("Action must be a non-empty string")
        
        return self._create_command(action, params or {})


class CommandBuilder:
    """Main command builder factory."""
    
    def actuator(self, module_id: str) -> ActuatorCommandBuilder:
        """Create an actuator command builder."""
        return ActuatorCommandBuilder(module_id)
    
    def display(self, module_id: str) -> DisplayCommandBuilder:
        """Create a display command builder."""
        return DisplayCommandBuilder(module_id)
    
    def mobile(self, module_id: str) -> MobileCommandBuilder:
        """Create a mobile command builder."""
        return MobileCommandBuilder(module_id)
    
    def sensor(self, module_id: str) -> SensorCommandBuilder:
        """Create a sensor command builder."""
        return SensorCommandBuilder(module_id)
    
    def custom(self, module_id: str) -> CustomCommandBuilder:
        """Create a custom command builder."""
        return CustomCommandBuilder(module_id)


def create_command_builder() -> CommandBuilder:
    """Create a new command builder instance."""
    return CommandBuilder()

File: libraries/typescript/src/builders.ts
import {
  QubiCommand,
  ModuleType,
  ServoParams,
  PositionParams,
  EyesParams,
  ExpressionParams,
  MovementParams,
  LocationParams,
  Expression,
} from './types';
import { QubiValidationError } from './errors';

export class CommandBuilder {
  actuator(moduleId: string): ActuatorCommandBuilder {
    return new ActuatorCommandBuilder(moduleId);
  }

  display(moduleId: string): DisplayCommandBuilder {
    return new DisplayCommandBuilder(moduleId);
  }

  mobile(moduleId: string): MobileCommandBuilder {
    return new MobileCommandBuilder(moduleId);
  }

  sensor(moduleId: string): SensorCommandBuilder {
    return new SensorCommandBuilder(moduleId);
  }

  custom(moduleId: string): CustomCommandBuilder {
    return new CustomCommandBuilder(moduleId);
  }
}

abstract class BaseCommandBuilder {
  constructor(
    protected readonly moduleId: string,
    protected readonly moduleType: ModuleType
  ) {}

  protected createCommand(action: string, params: Record<string, any>): QubiCommand {
    return {
      module_id: this.moduleId,
      module_type: this.moduleType,
      action,
      params,
    };
  }
}

export class ActuatorCommandBuilder extends BaseCommandBuilder {
  constructor(moduleId: string) {
    super(moduleId, 'actuator');
  }

  setServo(params: ServoParams): QubiCommand {
    this.validateServoParams(params);
    return this.createCommand('set_servo', params);
  }

  setPosition(params: PositionParams): QubiCommand {
    this.validatePositionParams(params);
    return this.createCommand('set_position', params);
  }

  getPosition(): QubiCommand {
    return this.createCommand('get_position', {});
  }

  stop(): QubiCommand {
    return this.createCommand('stop', {});
  }

  private validateServoParams(params: ServoParams): void {
    if (params.angle < 0 || params.angle > 180) {
      throw new QubiValidationError('Servo angle must be between 0 and 180 degrees');
    }

    if (params.speed !== undefined && (params.speed < 0 || params.speed > 255)) {
      throw new QubiValidationError('Servo speed must be between 0 and 255');
    }

    if (params.easing !== undefined) {
      const validEasing = ['linear', 'ease-in', 'ease-out'];
      if (!validEasing.includes(params.easing)) {
        throw new QubiValidationError(`Invalid easing type: ${params.easing}`);
      }
    }
  }

  private validatePositionParams(params: PositionParams): void {
    if (!Number.isFinite(params.x) || !Number.isFinite(params.y) || !Number.isFinite(params.z)) {
      throw new QubiValidationError('Position coordinates must be finite numbers');
    }
  }
}

export class DisplayCommandBuilder extends BaseCommandBuilder {
  constructor(moduleId: string) {
    super(moduleId, 'display');
  }

  setEyes(params: EyesParams): QubiCommand {
    this.validateEyesParams(params);
    return this.createCommand('set_eyes', params);
  }

  setExpression(params: ExpressionParams): QubiCommand {
    this.validateExpressionParams(params);
    return this.createCommand('set_expression', params);
  }

  clearDisplay(): QubiCommand {
    return this.createCommand('clear_display', {});
  }

  setBrightness(brightness: number): QubiCommand {
    if (brightness < 0 || brightness > 100) {
      throw new QubiValidationError('Brightness must be between 0 and 100');
    }
    return this.createCommand('set_brightness', { brightness });
  }

  private validateEyesParams(params: EyesParams): void {
    this.validateEyePosition(params.left_eye, 'left_eye');
    this.validateEyePosition(params.right_eye, 'right_eye');
  }

  private validateEyePosition(eye: { x: number; y: number }, eyeName: string): void {
    if (!Number.isInteger(eye.x) || !Number.isInteger(eye.y)) {
      throw new QubiValidationError(`${eyeName} coordinates must be integers`);
    }
    if (eye.x < 0 || eye.y < 0) {
      throw new QubiValidationError(`${eyeName} coordinates must be non-negative`);
    }
  }

  private validateExpressionParams(params: ExpressionParams): void {
    const validExpressions: Expression[] = ['happy', 'sad', 'surprised', 'neutral', 'angry'];
    if (!validExpressions.includes(params.expression)) {
      throw new QubiValidationError(`Invalid expression: ${params.expression}`);
    }

    if (params.intensity !== undefined && (params.intensity < 0 || params.intensity > 100)) {
      throw new QubiValidationError('Expression intensity must be between 0 and 100');
    }
  }
}

export class MobileCommandBuilder extends BaseCommandBuilder {
  constructor(moduleId: string) {
    super(moduleId, 'mobile');
  }

  move(params: MovementParams): QubiCommand {
    this.validateMovementParams(params);
    return this.createCommand('move', params);
  }

  setLocation(params: LocationParams): QubiCommand {
    this.validateLocationParams(params);
    return this.createCommand('set_location', params);
  }

  getLocation(): QubiCommand {
    return this.createCommand('get_location', {});
  }

  stop(): QubiCommand {
    return this.createCommand('stop', {});
  }

  rotate(angle: number, speed?: number): QubiCommand {
    if (!Number.isFinite(angle)) {
      throw new QubiValidationError('Rotation angle must be a finite number');
    }
    
    const params: Record<string, any> = { angle };
    if (speed !== undefined) {
      if (speed < 0 || speed > 100) {
        throw new QubiValidationError('Rotation speed must be between 0 and 100');
      }
      params.speed = speed;
    }
    
    return this.createCommand('rotate', params);
  }

  private validateMovementParams(params: MovementParams): void {
    if (!Number.isFinite(params.velocity)) {
      throw new QubiValidationError('Velocity must be a finite number');
    }

    if (!Number.isFinite(params.direction)) {
      throw new QubiValidationError('Direction must be a finite number');
    }

    if (params.duration !== undefined && (params.duration <= 0)) {
      throw new QubiValidationError('Duration must be positive');
    }
  }

  private validateLocationParams(params: LocationParams): void {
    if (!Number.isFinite(params.x) || !Number.isFinite(params.y)) {
      throw new QubiValidationError('Location coordinates must be finite numbers');
    }

    if (params.heading !== undefined && !Number.isFinite(params.heading)) {
      throw new QubiValidationError('Heading must be a finite number');
    }
  }
}

export class SensorCommandBuilder extends BaseCommandBuilder {
  constructor(moduleId: string) {
    super(moduleId, 'sensor');
  }

  read(sensorType?: string): QubiCommand {
    const params = sensorType ? { sensor_type: sensorType } : {};
    return this.createCommand('read', params);
  }

  startStreaming(sensorType: string, interval: number): QubiCommand {
    if (interval <= 0) {
      throw new QubiValidationError('Streaming interval must be positive');
    }
    
    return this.createCommand('start_streaming', {
      sensor_type: sensorType,
      interval,
    });
  }

  stopStreaming(sensorType?: string): QubiCommand {
    const params = sensorType ? { sensor_type: sensorType } : {};
    return this.createCommand('stop_streaming', params);
  }

  calibrate(sensorType: string): QubiCommand {
    return this.createCommand('calibrate', { sensor_type: sensorType });
  }

  getStatus(): QubiCommand {
    return this.createCommand('get_status', {});
  }
}

export class CustomCommandBuilder extends BaseCommandBuilder {
  constructor(moduleId: string) {
    super(moduleId, 'custom');
  }

  command(action: string, params: Record<string, any> = {}): QubiCommand {
    if (!action || typeof action !== 'string') {
      throw new QubiValidationError('Action must be a non-empty string');
    }
    
    return this.createCommand(action, params);
  }
}

// Convenience function to create a new command builder
export function createCommandBuilder(): CommandBuilder {
  return new CommandBuilder();
}